### PR TITLE
smt2: concat

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 29a0d2a5276e590dffb4cfe75694187dc25fd15f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94054538
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/solvers/smt2/smt2_parser.cpp [hunk 1]
```diff
@@ -612,15 +612,15 @@ exprt smt2_parsert::function_application()
...
+        // add the widths
+        std::size_t total_width = 0;
 
-        auto width0=to_unsignedbv_type(op[0].type()).get_width();
-        auto width1=to_unsignedbv_type(op[1].type()).get_width();
+        for(const auto &o : op)
+          total_width += to_unsignedbv_type(o.type()).get_width();
 
-        unsignedbv_typet t(width0+width1);
+        unsignedbv_typet t(total_width);
```
Comment (romainbrenguier) [inline src/solvers/smt2/smt2_parser.cpp]: const

File: src/solvers/smt2/smt2_parser.cpp [hunk 2]
```diff
@@ -612,15 +612,15 @@ exprt smt2_parsert::function_application()
...
       {
-        if(op.size()!=2)
-          throw error("concat takes two operands");
+        // add the widths
+        std::size_t total_width = 0;
 
-        auto width0=to_unsignedbv_type(op[0].type()).get_width();
-        auto width1=to_unsignedbv_type(op[1].type()).get_width();
+        for(const auto &o : op)
+          total_width += to_unsignedbv_type(o.type()).get_width();
```
Comment (romainbrenguier) [inline src/solvers/smt2/smt2_parser.cpp]: could maybe use `std::accumulate`

File: scripts/delete_failing_smt2_solver_tests [hunk 3]
```diff
@@ -88,8 +88,6 @@ rm equality_through_struct_containing_arrays2/test.desc
 rm equality_through_union1/test.desc
 rm equality_through_union2/test.desc
 rm equality_through_union3/test.desc
-rm gcc_bswap1/test.desc
```
Comment (romainbrenguier) [inline scripts/delete_failing_smt2_solver_tests:91]: Why are these tests deleted?

File: src/solvers/smt2/smt2_parser.cpp [hunk 4]
```diff
@@ -612,15 +612,15 @@ exprt smt2_parsert::function_application()
       }
       else if(id=="concat")
       {
-        if(op.size()!=2)
-          throw error("concat takes two operands");
+        // add the widths
```
Comment (romainbrenguier) [inline src/solvers/smt2/smt2_parser.cpp:618]: The commit message is not precisely describing the changes that are made

File: scripts/delete_failing_smt2_solver_tests (hunk 3 above)
Comment (romainbrenguier) [reply scripts/delete_failing_smt2_solver_tests:91]: Sorry, I now see this commit stops them from being deleted

**Comment by allredj** [comment]
Creating a test-gen bump.

**Comment by allredj** [comment]
Actually, no bump required. Automatic bump should pass once this PR is rebased onto cbmc/develop.

**Review by allredj — APPROVED** [review]

File: src/solvers/smt2/smt2_parser.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: ok

File: src/solvers/smt2/smt2_parser.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: ok

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 24883f024a5321a31f565eadda07470726d171d9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94364406

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**